Add toggle to suppress KeyOrder element in maps

Mjolnir42 opened the pull request:
Introduce UclExportKeyOrder variable that allows to
disable adding the KeyOrder element to maps.

This allows to avoid errors when encoding/json.Unmarshal
is used to convert the map[string]interface{} output to
struct and the struct contains a map[string]string element.
Otherwise the following error is returned:

  json: cannot unmarshal array into Go value of type string

The KeyOrder element is exported by default to preserve
behaviour.

Mjolnir42 committed:
Add toggle to suppress KeyOrder element in maps

Introduce UclExportKeyOrder variable that allows to
disable adding the KeyOrder element to maps.

This allows to avoid errors when encoding/json.Unmarshal
is used to convert the map[string]interface{} output to
struct and the struct contains a map[string]string element.
Otherwise the following error is returned:

  json: cannot unmarshal array into Go value of type string

The KeyOrder element is exported by default to preserve
behaviour.

l-d-x commented:
That's a good feature to have. Although it might be better expanded to block out the allocation of korder (lines 267-275) so that it doesn't need to be done for every tag that is parsed. If you can prepare another patch, I'll re-review it. Thanks.

Mjolnir42 committed:
Only build KeyOrder array if it is to be exported

Change UclExportKeyOrder behaviour from only not setting the
KeyOrder element of the map to only initialize it if it is to
be exported.

Then update the KeyOrder conditional on it being initialized, so
the toggle is only evaluated in one code location.

Requested by: Leon Dang

Mjolnir42 commented:
I have updated it, leaving the codepath for having the keyorder as intact as I could manage. Please have a look at it.
